### PR TITLE
SNOW-670956 Support qualify

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Support `DataFrame.qualify`, which corresponds to [QUALIFY](https://docs.snowflake.com/en/sql-reference/constructs/qualify) clause in Snowflake SQL.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 1]
```diff
@@ -394,12 +394,14 @@ def __init__(
         order_by: Optional[List[Expression]] = None,
         limit_: Optional[int] = None,
         offset: Optional[int] = None,
+        qualify: Optional[Expression] = None,
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:397]: It's interesting that we have `qualify` here but not `having` 🤔

**Review by sfc-gh-sfan — COMMENTED** [review]
What is the behavior when a dataframe has no window function and tries to use qualify? Given that we don't have a check in client, I guess it would be the server to raise this error. Is the error easier to comprehend for a snowpark user? Or should we actually add some client side validation based on class type (because we know if certain construct is a window function)?

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py (hunk 1 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:397]: True. There is no active demand for these SQL features that don't have a direct counterpart in Spark. It shouldn't be much work to support it tho.

**Comment by sfc-gh-stan** [comment]
> What is the behavior when a dataframe has no window function and tries to use qualify? Given that we don't have a check in client, I guess it would be the server to raise this error. Is the error easier to comprehend for a snowpark user? Or should we actually add some client side validation based on class type (because we know if certain construct is a window function)?

Good question. Server will raise a SQL compilation exception that looks something like `SQL compilation error: error line 1 at position 228
found QUALIFY clause but no window function`. I think this is pretty comprehensible, we could do client side verification but only partially since the use case where `df.select(<window_expr>.alias("criteria")).qualify(col("criteria") > 1)` cannot be verified easily (we never check logical plan of the subquery).

File: src/snowflake/snowpark/dataframe.py [hunk 2]
```diff
@@ -3496,6 +3497,41 @@ def with_column_renamed(self, existing: ColumnOrName, new: str) -> "DataFrame":
         ]
         return self.select(new_columns)
 
+    def qualify(self, expr: ColumnOrSqlExpr) -> "DataFrame":
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/dataframe.py:3500]: Just to brainstorm, is it possible to still use api `filter` and generate SQL with `qualify` when window functions are used?

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/dataframe.py:3500]: This is similar to Shixuan's question. To fully support that, we need to detect window expression in the current dataframe's projection, so for something like `df.select(window_expr.alias("winwin").filter(col("winwin") > 12)` we need to traverse the dataframe's logical plan. It's probably easier for simplified paths.

**Pull request closed without merging**